# DML and DMR Analysis

In this notebook, I will examine the location of differentially methylated loci (DML) and regions (DMR) in the *C. virginica* genome. The DML and DMR were identified using methylKit in [this R script](https://github.com/fish546-2018/yaamini-virginica/tree/master/analyses/2018-10-25-MethylKit).

Methods:

0. Prepare for Analyses
1. Locate Files and Set Variable Paths
2. Identify Overlaps with Genomic Feature Tracks
3. Identify Overlaps between CG motifs and other Feature Tracks
4. Identify Overlaps between Transposable Elements and Other Feature Tracks
5. Calculate Overlap Proprtions
6. Gene Flanking

## 0. Prepare for Analyses

### 0a. Set Working Directory

In [1]:
pwd

'/Users/yaamini/Documents/yaamini-virginica/notebooks'

In [2]:
cd ../analyses/

/Users/yaamini/Documents/yaamini-virginica/analyses


In [3]:
pwd

'/Users/yaamini/Documents/yaamini-virginica/analyses'

In [1]:
!mkdir 2018-11-01-DML-and-DMR-Analysis

In [4]:
ls -F

2018-10-25-MethylKit/                2019-01-15-Sample-Clustering/
2018-11-01-DML-and-DMR-Analysis/     2019-03-07-IGV-Verification/
2018-12-02-Gene-Enrichment-Analysis/ README.md


In [3]:
cd 2018-11-01-DML-and-DMR-Analysis/

/Users/yaamini/Documents/yaamini-virginica/analyses/2018-11-01-DML-and-DMR-Analysis


### 0b. Download Genome Feature Files

I will be using the following Genome Feature Tracks [Roberts Lab Genomic Resources wiki page](https://github.com/RobertsLab/resources/wiki/Genomic-Resources):

1. Exon: Coding regions
2. Intron: Regions that are removed
3. mRNA: Code for proteins! The mRNA track includes both introns and exons
4. Transposable elements (all): Transposable elements located using information all species in the RepeatMasker databse (see [Sam's notes](http://onsnetwork.org/kubu4/2018/08/28/transposable-element-mapping-crassostrea-virginica-genome-cvirginica_v300-using-repeatmasker-4-07/) for more information)
5. Tranpsosable elements (_C. gigas_): Transposable elements located using information from _C. gigas_ only (see [Sam's notes](http://onsnetwork.org/kubu4/2018/08/28/transposable-element-mapping-crassostrea-virginica-genome-cvirginica_v300-using-repeatmasker-4-07/) for more information)
4. CG motifs: Regions with CGs where methylation can occur

In [6]:
!curl http://eagle.fish.washington.edu/Cvirg_tracks/C_virginica-3.0_Gnomon_exon.bed > C_virginica-3.0_Gnomon_exon.bed

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 20.7M  100 20.7M    0     0  45.7M      0 --:--:-- --:--:-- --:--:-- 47.4M


In [7]:
!curl http://eagle.fish.washington.edu/Cvirg_tracks/C_virginica-3.0_intron.bed > C_virginica-3.0_intron.bed

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 9260k  100 9260k    0     0  44.1M      0 --:--:-- --:--:-- --:--:-- 44.9M


In [8]:
!curl http://eagle.fish.washington.edu/Cvirg_tracks/C_virginica-3.0_Gnomon_mRNA.gff3 > C_virginica-3.0_Gnomon_mRNA.gff3

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 26.4M  100 26.4M    0     0  53.0M      0 --:--:-- --:--:-- --:--:-- 53.2M


In [9]:
!curl http://owl.fish.washington.edu/halfshell/genomic-databank/C_virginica-3.0_TE-all.gff > C_virginica-3.0_TE-all.gff

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 63.0M  100 63.0M    0     0  45.2M      0  0:00:01  0:00:01 --:--:-- 45.3M


In [10]:
!curl http://owl.fish.washington.edu/halfshell/genomic-databank/C_virginica-3.0_TE-Cg.gff > C_virginica-3.0_TE-Cg.gff

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 57.4M  100 57.4M    0     0  47.4M      0  0:00:01  0:00:01 --:--:-- 47.5M


In [11]:
!curl http://eagle.fish.washington.edu/Cvirg_tracks/C_virginica-3.0_CG-motif.bed > C_virginica-3.0_CG-motif.bed

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  533M  100  533M    0     0  53.1M      0  0:00:10  0:00:10 --:--:-- 54.6M


In [14]:
!ls -F

2018-11-07-DML-Exon.txt
2018-11-07-DML-Intron.txt
2018-11-07-DML-TE-Cg.txt
2018-11-07-DML-TE-all.txt
2018-11-07-DML-mRNA.txt
2018-11-07-DMR-Exon.txt
2018-11-07-DMR-Intron.txt
2018-11-07-DMR-TE-Cg.txt
2018-11-07-DMR-TE-all.txt
2018-11-07-DMR-mRNA.txt
2018-11-07-Exon-CGmotif.txt
2018-11-07-Exon-TE-Cv.txt
2018-11-07-Exon-TE-all.txt
2018-11-07-Intron-CGmotif.txt
2018-11-07-Intron-TE-Cv.txt
2018-11-07-Intron-TE-all.txt
2018-11-07-TE-Cv-CGmotif.txt
2018-11-07-TE-all-CGmotif.txt
2018-11-07-Unique-Genes-in-DML-mRNA-Overlap.txt
2018-11-07-Unique-Genes-in-DMR-mRNA-Overlap.txt
2018-11-07-mRNA-CGmotif.txt
2018-11-07-mRNA-TE-Cv.txt
2018-11-07-mRNA-TE-all.txt
C_virginica-3.0_CG-motif.bed
C_virginica-3.0_Gnomon_exon.bed
C_virginica-3.0_Gnomon_mRNA.gff3
C_virginica-3.0_TE-Cg.gff
C_virginica-3.0_TE-all.gff
C_virginica-3.0_intron.bed


## 1. Locate Relevant Files and Set Variable Path Names

### 1a. Set Variable Path Names

Setting the variable path names allows me to reuse this script with different input files or different paths to programs without manually changing the file names each time.

In [4]:
bedtoolsDirectory = "/Users/Shared/bioinformatics/bedtools2/bin/"

In [5]:
DMLlist = "../../analyses/2018-10-25-MethylKit/2019-04-05-DML-Destrand-5x-Locations.bed"

In [8]:
DMRlist = "../../analyses/2018-10-25-MethylKit/2018-11-07-DMR-Locations.bed"

In [6]:
exonList = "C_virginica-3.0_Gnomon_exon.bed"

In [7]:
intronList = "C_virginica-3.0_intron.bed"

In [9]:
mRNAList = "C_virginica-3.0_Gnomon_mRNA.gff3"

In [8]:
transposableElementsAll = "C_virginica-3.0_TE-all.gff"

In [11]:
transposableElementsCg = "C_virginica-3.0_TE-Cg.gff"

In [9]:
CGMotifList = "C_virginica-3.0_CG-motif.bed"

### 1b. Confirm Variable Path Works and Characterize Files

The BEDfiles with DML and DMR can be viewed below. Columns are are the chromosome, start position, end position, strand, and fold difference with direction. The files only have DML and DMR that were at least 50% different between the two treatments (control and elevated pCO<sub>2</sub>).

In [14]:
#Previewing the files
!head {DMLlist}

NC_035780.1	571138	571140	58
NC_035780.1	1882691	1882693	64
NC_035780.1	1885022	1885024	61
NC_035780.1	1933499	1933501	51
NC_035780.1	1958998	1959000	50
NC_035780.1	2538924	2538926	-50
NC_035780.1	2541726	2541728	-54
NC_035780.1	2584492	2584494	56
NC_035780.1	2586508	2586510	-53
NC_035780.1	2588794	2588796	-53


In [13]:
#Counting the number of lines to count DML
!wc -l {DMLlist}

     598 ../../analyses/2018-10-25-MethylKit/2019-04-05-DML-Destrand-5x-Locations.bed


In [17]:
!head {DMRlist}

NC_035780.1	571100	571200	DMR	58
NC_035780.1	573700	573800	DMR	52
NC_035780.1	1885000	1885100	DMR	51
NC_035780.1	1933500	1933600	DMR	53
NC_035780.1	4285700	4285800	DMR	-51
NC_035780.1	24159600	24159700	DMR	51
NC_035780.1	26629500	26629600	DMR	65
NC_035780.1	28563400	28563500	DMR	59
NC_035780.1	29883000	29883100	DMR	-55
NC_035780.1	31302900	31303000	DMR	-61


In [18]:
#Counting the number of DMR
!wc -l {DMRlist}

     162 ../../analyses/2018-10-25-MethylKit/2018-11-07-DMR-Locations.bed


In [19]:
!head {exonList}

NC_035780.1	13578	13603
NC_035780.1	14237	14290
NC_035780.1	14557	14594
NC_035780.1	28961	29073
NC_035780.1	30524	31557
NC_035780.1	31736	31887
NC_035780.1	31977	32565
NC_035780.1	32959	33324
NC_035780.1	66869	66897
NC_035780.1	64123	64334


In [20]:
!wc -l {exonList}

  731279 C_virginica-3.0_Gnomon_exon.bed


In [21]:
!head {intronList}

NC_035780.1	28961	28961
NC_035780.1	29074	30524
NC_035780.1	31558	31736
NC_035780.1	31888	31977
NC_035780.1	32566	32959
NC_035780.1	43110	43112
NC_035780.1	44359	45913
NC_035780.1	46507	64123
NC_035780.1	64335	66869
NC_035780.1	85606	85606


In [22]:
!wc -l {intronList}

  319262 C_virginica-3.0_intron.bed


In [60]:
!head -n 1 {mRNAList}

NC_035780.1	Gnomon	mRNA	28961	33324	.	+	.	ID=rna1;Parent=gene1;Dbxref=GeneID:111126949,Genbank:XM_022471938.1;Name=XM_022471938.1;gbkey=mRNA;gene=LOC111126949;model_evidence=Supporting evidence includes similarity to: 3 Proteins%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 21 samples with support for all annotated introns;product=UNC5C-like protein;transcript_id=XM_022471938.1


In [24]:
!wc -l {mRNAList}

   60201 C_virginica-3.0_Gnomon_mRNA.gff3


In [27]:
!head {transposableElementsAll}

##gff-version 2
##date 2018-08-23
##sequence-region Cvirginica_v300.fa
NC_007175.2	RepeatMasker	similarity	262	1389	31.1	+	.	Target "Motif:REP-6_LMi" 2920 4055
NC_007175.2	RepeatMasker	similarity	1728	1947	26.1	-	.	Target "Motif:REP-6_LMi" 14320 14534
NC_007175.2	RepeatMasker	similarity	1866	2013	33.6	+	.	Target "Motif:LSU-rRNA_Cel" 2372 2520
NC_007175.2	RepeatMasker	similarity	2129	2367	20.5	-	.	Target "Motif:REP-6_LMi" 13886 14118
NC_007175.2	RepeatMasker	similarity	2836	2980	31.5	+	.	Target "Motif:REP-6_LMi" 6216 6359
NC_007175.2	RepeatMasker	similarity	3196	3277	30.5	+	.	Target "Motif:REP-6_LMi" 6572 6653
NC_007175.2	RepeatMasker	similarity	5168	5532	32.9	+	.	Target "Motif:REP-6_LMi" 4620 4983


In [29]:
!wc -l {transposableElementsAll}

  692371 C_virginica-3,0_TE-all.gff


In [30]:
!head {transposableElementsCg}

##gff-version 2
##date 2018-08-27
##sequence-region Cvirginica_v300.fa
NC_007175.2	RepeatMasker	similarity	1866	2013	33.6	+	.	Target "Motif:LSU-rRNA_Cel" 2372 2520
NC_007175.2	RepeatMasker	similarity	6529	6628	19.0	+	.	Target "Motif:(TA)n" 2 102
NC_035780.1	RepeatMasker	similarity	1473	1535	 0.0	+	.	Target "Motif:(TAACCC)n" 1 63
NC_035780.1	RepeatMasker	similarity	5080	7289	32.5	-	.	Target "Motif:Gypsy-62_CGi-I" 2102 4631
NC_035780.1	RepeatMasker	similarity	7423	7489	25.4	-	.	Target "Motif:Gypsy-62_CGi-I" 2097 2163
NC_035780.1	RepeatMasker	similarity	7623	8079	34.1	-	.	Target "Motif:Gypsy-62_CGi-I" 1516 1975
NC_035780.1	RepeatMasker	similarity	8261	8295	14.1	+	.	Target "Motif:(CTCCT)n" 1 33


In [31]:
!wc -l {transposableElementsCg}

  626665 C_virginica-3.0_TE-Cg.gff


In [25]:
!head {CGMotifList}

NC_035780.1	28	30	CG_motif
NC_035780.1	54	56	CG_motif
NC_035780.1	75	77	CG_motif
NC_035780.1	93	95	CG_motif
NC_035780.1	103	105	CG_motif
NC_035780.1	116	118	CG_motif
NC_035780.1	134	136	CG_motif
NC_035780.1	159	161	CG_motif
NC_035780.1	209	211	CG_motif
NC_035780.1	224	226	CG_motif


In [26]:
!wc -l {CGMotifList}

 14458703 C_virginica-3.0_CG-motif.bed


## 2. Identify DML and DMR Overlaps with Genomic Feature Tracks

To identify the location of DML and DMR in the *C. virginica* genome, I will use `intersect` from `bedtools`. [The BEDtools suite](http://bedtools.readthedocs.io/en/latest/content/bedtools-suite.html) allows me to easily find overlapping regions of different bed files.

In [36]:
! {bedtoolsDirectory}intersectBed -h


Tool:    bedtools intersect (aka intersectBed)
Version: v2.26.0
Summary: Report overlaps between two feature files.

Usage:   bedtools intersect [OPTIONS] -a <bed/gff/vcf/bam> -b <bed/gff/vcf/bam>

	Note: -b may be followed with multiple databases and/or 
	wildcard (*) character(s). 
Options: 
	-wa	Write the original entry in A for each overlap.

	-wb	Write the original entry in B for each overlap.
		- Useful for knowing _what_ A overlaps. Restricted by -f and -r.

	-loj	Perform a "left outer join". That is, for each feature in A
		report each overlap with B.  If no overlaps are found, 
		report a NULL feature for B.

	-wo	Write the original A and B entries plus the number of base
		pairs of overlap between the two features.
		- Overlaps restricted by -f and -r.
		  Only A features with overlap are reported.

	-wao	Write the original A and B entries plus the number of base
		pairs of overlap between the two features.
		- Overlapping features restricted by -f and -r.
		  However, A fea

### 2a. Exons

#### DML

In [14]:
! {bedtoolsDirectory}intersectBed \
-u \
-a {DMLlist} \
-b {exonList} \
| wc -l
!echo "DML overlaps with exons"

     368
DML overlaps with exons


In [15]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a {DMLlist} \
-b {exonList} \
> 2018-11-07-DML-Exon.txt

In [16]:
!head 2018-11-07-DML-Exon.txt

NC_035780.1	571138	571140	58	NC_035780.1	570942	571194
NC_035780.1	2538924	2538926	-50	NC_035780.1	2538624	2538955
NC_035780.1	2586508	2586510	-53	NC_035780.1	2586438	2586557
NC_035780.1	2589720	2589722	57	NC_035780.1	2589716	2589955
NC_035780.1	4286286	4286288	67	NC_035780.1	4286174	4286407
NC_035780.1	4286802	4286804	-62	NC_035780.1	4286783	4286927
NC_035780.1	4289628	4289630	-52	NC_035780.1	4288592	4290756
NC_035780.1	8693287	8693289	-52	NC_035780.1	8692509	8693320
NC_035780.1	9110274	9110276	-63	NC_035780.1	9109982	9111843
NC_035780.1	12631453	12631455	60	NC_035780.1	12630576	12631487


#### DMR

In [35]:
! {bedtoolsDirectory}intersectBed \
-u \
-a {DMRlist} \
-b {exonList} \
| wc -l
!echo "DMR overlaps with exons"

      64
DMR overlaps with exons


In [36]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a {DMRlist} \
-b {exonList} \
> 2018-11-07-DMR-Exon.txt

In [37]:
!head 2018-11-07-DMR-Exon.txt

NC_035780.1	571100	571194	DMR	58	NC_035780.1	570942	571194
NC_035780.1	573700	573800	DMR	52	NC_035780.1	573630	573906
NC_035780.1	573700	573800	DMR	52	NC_035780.1	573630	573906
NC_035780.1	1933574	1933600	DMR	53	NC_035780.1	1933574	1933615
NC_035780.1	47335000	47335100	DMR	-66	NC_035780.1	47334080	47336192
NC_035780.1	48394800	48394900	DMR	-63	NC_035780.1	48394159	48395287
NC_035780.1	61138200	61138300	DMR	-79	NC_035780.1	61138000	61140417
NC_035781.1	6831300	6831302	DMR	-59	NC_035781.1	6831093	6831302
NC_035781.1	6831300	6831302	DMR	-59	NC_035781.1	6831093	6831302
NC_035781.1	6831300	6831302	DMR	-59	NC_035781.1	6831093	6831302


### 2b. Introns

#### DML

In [17]:
! {bedtoolsDirectory}intersectBed \
-u \
-a {DMLlist} \
-b {intronList} \
| wc -l
!echo "DML overlaps with introns"

     191
DML overlaps with introns


In [18]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a {DMLlist} \
-b {intronList} \
> 2018-11-07-DML-Intron.txt

In [19]:
!head 2018-11-07-DML-Intron.txt

NC_035780.1	401630	401632	53	NC_035780.1	401605	401801
NC_035780.1	1882691	1882693	64	NC_035780.1	1882356	1882972
NC_035780.1	1885022	1885024	61	NC_035780.1	1884755	1886043
NC_035780.1	1933499	1933501	51	NC_035780.1	1932877	1933574
NC_035780.1	2541726	2541728	-54	NC_035780.1	2538956	2541769
NC_035780.1	2584492	2584494	56	NC_035780.1	2584154	2584505
NC_035780.1	4288213	4288215	-58	NC_035780.1	4288129	4288231
NC_035780.1	8833124	8833126	60	NC_035780.1	8832172	8833700
NC_035780.1	15414934	15414935	51	NC_035780.1	15414677	15414935
NC_035780.1	17488958	17488960	-57	NC_035780.1	17488943	17489179


#### DMR

In [41]:
! {bedtoolsDirectory}intersectBed \
-u \
-a {DMRlist} \
-b {intronList} \
| wc -l
!echo "DMR overlaps with introns"

     112
DMR overlaps with introns


In [42]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a {DMRlist} \
-b {intronList} \
> 2018-11-07-DMR-Intron.txt

In [43]:
!head 2018-11-07-DMR-Intron.txt

NC_035780.1	571195	571200	DMR	58	NC_035780.1	571195	572677
NC_035780.1	1885000	1885100	DMR	51	NC_035780.1	1884755	1886043
NC_035780.1	1933500	1933574	DMR	53	NC_035780.1	1932877	1933574
NC_035780.1	4285700	4285800	DMR	-51	NC_035780.1	4285382	4285831
NC_035780.1	26629500	26629600	DMR	65	NC_035780.1	26621919	26632637
NC_035780.1	28563400	28563500	DMR	59	NC_035780.1	28563400	28564616
NC_035780.1	29883000	29883100	DMR	-55	NC_035780.1	29882984	29883643
NC_035780.1	31302900	31303000	DMR	-61	NC_035780.1	31302842	31303152
NC_035780.1	31303300	31303400	DMR	-53	NC_035780.1	31303293	31303603
NC_035780.1	33209900	33210000	DMR	-54	NC_035780.1	33209785	33210978


### 2c. mRNA

#### DML

In [20]:
! {bedtoolsDirectory}intersectBed \
-u \
-a {DMLlist} \
-b {mRNAList} \
| wc -l
!echo "DML overlaps with mRNA"

     549
DML overlaps with mRNA


In [21]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a {DMLlist} \
-b {mRNAList} \
> 2018-11-07-DML-mRNA.txt

In [22]:
!head -n 1 2018-11-07-DML-mRNA.txt

NC_035780.1	401630	401632	53	NC_035780.1	Gnomon	mRNA	394983	409280	.	-	.	ID=rna37;Parent=gene27;Dbxref=GeneID:111117672,Genbank:XM_022456873.1;Name=XM_022456873.1;Note=The sequence of the model RefSeq transcript was modified relative to this genomic sequence to represent the inferred CDS: inserted 1 base in 1 codon;exception=unclassified transcription discrepancy;gbkey=mRNA;gene=LOC111117672;model_evidence=Supporting evidence includes similarity to: 6 Proteins%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 24 samples with support for all annotated introns;product=Niemann-Pick C1 protein-like;transcript_id=XM_022456873.1


I know how many overlaps there are, but I also want to know how many unique genes have DMLs in them. For this, I will use the following code:

`cut -f13 2018-11-07-DML-mRNA.txt | sort | uniq -c`

`cut` is the command that isolates the column information. The column is piped into `sort`, then that output is counted for unique lines by `uniq`. I will save the output from this command as a new file.

In [23]:
! cut -f13 2018-11-07-DML-mRNA.txt | sort | uniq -c > 2018-11-07-Unique-Genes-in-DML-mRNA-Overlap.txt

In [24]:
!head 2018-11-07-Unique-Genes-in-DML-mRNA-Overlap.txt

   1 ID=rna10013;Parent=gene5874;Dbxref=GeneID:111118920,Genbank:XM_022458635.1;Name=XM_022458635.1;gbkey=mRNA;gene=LOC111118920;model_evidence=Supporting evidence includes similarity to: 1 Protein%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 9 samples with support for all annotated introns;product=structural maintenance of chromosomes protein 6-like%2C transcript variant X1;transcript_id=XM_022458635.1
   1 ID=rna10014;Parent=gene5874;Dbxref=GeneID:111118920,Genbank:XM_022458636.1;Name=XM_022458636.1;gbkey=mRNA;gene=LOC111118920;model_evidence=Supporting evidence includes similarity to: 2 Proteins%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 9 samples with support for all annotated introns;product=structural maintenance of chromosomes protein 6-like%2C transcript variant X2;transcript_id=XM_022458636.1
   1 ID=rna10055;Parent=gene5904;Dbxref=GeneID:111121117,Genbank:XM_022462246.1;Name=XM_0224622

In [25]:
!wc -l 2018-11-07-Unique-Genes-in-DML-mRNA-Overlap.txt

    1430 2018-11-07-Unique-Genes-in-DML-mRNA-Overlap.txt


The DMLs overlap with 1430 unique genes.

#### DMR

In [50]:
! {bedtoolsDirectory}intersectBed \
-u \
-a {DMRlist} \
-b {mRNAList} \
| wc -l
!echo "DMR overlaps with mRNA"

     139
DMR overlaps with mRNA


In [51]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a {DMRlist} \
-b {mRNAList} \
> 2018-11-07-DMR-mRNA.txt

In [52]:
!head 2018-11-07-DMR-mRNA.txt

NC_035780.1	571100	571200	DMR	58	NC_035780.1	Gnomon	mRNA	544088	573497	.	+	.	ID=rna48;Parent=gene35;Dbxref=GeneID:111114201,Genbank:XM_022452489.1;Name=XM_022452489.1;Note=The sequence of the model RefSeq transcript was modified relative to this genomic sequence to represent the inferred CDS: inserted 2 bases in 2 codons;exception=unclassified transcription discrepancy;gbkey=mRNA;gene=LOC111114201;model_evidence=Supporting evidence includes similarity to: 4 Proteins%2C and 99%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 9 samples with support for all annotated introns;product=vacuolar protein sorting-associated protein 13B-like;transcript_id=XM_022452489.1
NC_035780.1	573700	573800	DMR	52	NC_035780.1	Gnomon	mRNA	573630	585444	.	-	.	ID=rna49;Parent=gene36;Dbxref=GeneID:111114212,Genbank:XM_022452506.1;Name=XM_022452506.1;gbkey=mRNA;gene=LOC111114212;model_evidence=Supporting evidence includes similarity to: 4 Proteins%2C and 100%25 coverage of the annot

In [53]:
! cut -f14 2018-11-07-DMR-mRNA.txt | sort | uniq -c > 2018-11-07-Unique-Genes-in-DMR-mRNA-Overlap.txt

In [54]:
!head 2018-11-07-Unique-Genes-in-DMR-mRNA-Overlap.txt

   1 ID=rna10182;Parent=gene5989;Dbxref=GeneID:111121400,Genbank:XM_022462674.1;Name=XM_022462674.1;gbkey=mRNA;gene=LOC111121400;model_evidence=Supporting evidence includes similarity to: 3 Proteins%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 16 samples with support for all annotated introns;product=DNA excision repair protein ERCC-6-like;transcript_id=XM_022462674.1
   1 ID=rna10216;Parent=gene6005;Dbxref=GeneID:111120829,Genbank:XM_022461817.1;Name=XM_022461817.1;gbkey=mRNA;gene=LOC111120829;model_evidence=Supporting evidence includes similarity to: 1 Protein%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 23 samples with support for all annotated introns;product=serine/arginine-rich splicing factor 7-like%2C transcript variant X1;transcript_id=XM_022461817.1
   1 ID=rna10452;Parent=gene6155;Dbxref=GeneID:111118143,Genbank:XM_022457464.1;Name=XM_022457464.1;Note=The sequence of the model RefSeq tr

In [55]:
!wc -l 2018-11-07-Unique-Genes-in-DMR-mRNA-Overlap.txt

     305 2018-11-07-Unique-Genes-in-DMR-mRNA-Overlap.txt


The DMRs overlap with 305 unique genes.

### 2c. Transposable Elements (All)

#### DML

In [26]:
! {bedtoolsDirectory}intersectBed \
-u \
-a {DMLlist} \
-b {transposableElementsAll} \
| wc -l
!echo "DML overlaps with transposable elements (all)"

      57
DML overlaps with transposable elements (all)


In [27]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a {DMLlist} \
-b {transposableElementsAll} \
> 2018-11-07-DML-TE-all.txt

In [28]:
!head 2018-11-07-DML-TE-all.txt

NC_035780.1	8833124	8833126	60	NC_035780.1	RepeatMasker	similarity	8833042	8833288	18.2	-	.	Target "Motif:CVA" 1 272
NC_035780.1	22177828	22177830	-51	NC_035780.1	RepeatMasker	similarity	22177766	22177877	22.3	-	.	Target "Motif:DNA9-6_CGi" 1 115
NC_035780.1	57337100	57337102	-54	NC_035780.1	RepeatMasker	similarity	57337042	57337128	18.6	-	.	Target "Motif:DNA2-2_CGi" 413 498
NC_035780.1	58135767	58135769	74	NC_035780.1	RepeatMasker	similarity	58135699	58135837	22.4	+	.	Target "Motif:BivaMD-SINE1_CrVi" 169 314
NC_035781.1	22439769	22439771	53	NC_035781.1	RepeatMasker	similarity	22439740	22439796	28.1	+	.	Target "Motif:Mariner-6_AMi" 698 754
NC_035781.1	29178318	29178320	-55	NC_035781.1	RepeatMasker	similarity	29177336	29178341	16.0	-	.	Target "Motif:CVA" 2 863
NC_035781.1	54151548	54151550	54	NC_035781.1	RepeatMasker	similarity	54150482	54151750	14.3	+	.	Target "Motif:CVA" 1 1018
NC_035781.1	59742649	59742651	-65	NC_035781.1	RepeatMasker	similarity	59742603	59742651	 4.2	+	.	Target "Moti

#### DMR

In [59]:
! {bedtoolsDirectory}intersectBed \
-u \
-a {DMRlist} \
-b {transposableElementsAll} \
| wc -l
!echo "DMR overlaps with transposable elements (all)"

      39
DMR overlaps with transposable elements (all)


In [138]:
! {bedtoolsDirectory}intersectBed \
-u \
-a {DMRlist} \
-b {transposableElementsAll} \
> 2018-11-07-DMR-TE-all.txt

In [139]:
!head 2018-11-07-DMR-TE-all.txt

NC_035780.1	24159600	24159700	DMR	51
NC_035780.1	29883000	29883100	DMR	-55
NC_035780.1	33945900	33946000	DMR	-51
NC_035780.1	46044700	46044800	DMR	-53
NC_035780.1	47335000	47335100	DMR	-66
NC_035781.1	30962200	30962300	DMR	63
NC_035781.1	51566900	51567000	DMR	-61
NC_035781.1	54151500	54151600	DMR	55
NC_035782.1	2787300	2787400	DMR	-53
NC_035782.1	7518400	7518500	DMR	64


### 2e. Transposable Elements (_C. gigas_ only)

#### DML

In [29]:
! {bedtoolsDirectory}intersectBed \
-u \
-a {DMLlist} \
-b {transposableElementsCg} \
| wc -l
!echo "DML overlaps with transposable elements (Cg)"

      39
DML overlaps with transposable elements (Cg)


In [30]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a {DMLlist} \
-b {transposableElementsCg} \
> 2018-11-07-DML-TE-Cg.txt

In [31]:
!head 2018-11-07-DML-TE-Cg.txt

NC_035780.1	8833124	8833126	60	NC_035780.1	RepeatMasker	similarity	8833045	8833287	22.6	-	.	Target "Motif:Helitron-N2f_CGi" 1 276
NC_035780.1	22177828	22177830	-51	NC_035780.1	RepeatMasker	similarity	22177766	22177877	22.3	-	.	Target "Motif:DNA9-6_CGi" 1 115
NC_035780.1	57337100	57337102	-54	NC_035780.1	RepeatMasker	similarity	57337042	57337128	18.6	-	.	Target "Motif:DNA2-2_CGi" 413 498
NC_035781.1	29178318	29178320	-55	NC_035781.1	RepeatMasker	similarity	29177333	29178341	24.4	-	.	Target "Motif:Helitron-N2d_CGi" 2 863
NC_035781.1	54151548	54151550	54	NC_035781.1	RepeatMasker	similarity	54150483	54151741	23.3	+	.	Target "Motif:Helitron-N2f_CGi" 1 1018
NC_035781.1	59742649	59742651	-65	NC_035781.1	RepeatMasker	similarity	59742603	59742651	 4.2	+	.	Target "Motif:(ACTAACG)n" 1 49
NC_035782.1	34498893	34498895	-55	NC_035782.1	RepeatMasker	similarity	34498501	34500091	24.8	+	.	Target "Motif:Helitron-N40_CGi" 1 1569
NC_035782.1	34498895	34498897	-71	NC_035782.1	RepeatMasker	similarity	344985

#### DMR

In [22]:
! {bedtoolsDirectory}intersectBed \
-u \
-a {DMRlist} \
-b {transposableElementsCg} \
| wc -l
!echo "DMR overlaps with transposable elements (Cg)"

      23
DMR overlaps with transposable elements (Cg)


In [66]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a {DMRlist} \
-b {transposableElementsCg} \
> 2018-11-07-DMR-TE-Cg.txt

In [67]:
!head 2018-11-07-DMR-TE-Cg.txt

NC_035780.1	29883075	29883100	DMR	-55	NC_035780.1	RepeatMasker	similarity	29883076	29883262	25.4	+	.	Target "Motif:DNA9-4_CGi" 776 964
NC_035780.1	33945935	33945973	DMR	-51	NC_035780.1	RepeatMasker	similarity	33945936	33945973	 5.3	+	.	Target "Motif:DNA9-5_CGi" 1 38
NC_035780.1	33945972	33946000	DMR	-51	NC_035780.1	RepeatMasker	similarity	33945973	33946026	 3.7	+	.	Target "Motif:DNA8-12_CGi" 645 699
NC_035781.1	30962200	30962300	DMR	63	NC_035781.1	RepeatMasker	similarity	30962179	30962531	18.4	+	.	Target "Motif:Helitron-N2d_CGi" 3 346
NC_035781.1	51566900	51566921	DMR	-61	NC_035781.1	RepeatMasker	similarity	51566285	51566921	25.2	-	.	Target "Motif:Sola3-1_CGi" 3273 3922
NC_035781.1	51566900	51567000	DMR	-61	NC_035781.1	RepeatMasker	similarity	51566310	51567155	27.8	-	.	Target "Motif:Helitron-N43_CGi" 1 1221
NC_035781.1	54151500	54151600	DMR	55	NC_035781.1	RepeatMasker	similarity	54150483	54151741	23.3	+	.	Target "Motif:Helitron-N2f_CGi" 1 1018
NC_035782.1	2787300	2787318	DMR	-53	NC_035

## 3. Identify Overlaps between CG Motif and Other Genome Feature Tracks

It's also useful to understand where the CG regions are in relation to exons, introns, mRNA, and transposable elements!

### 3a. Exons

In [68]:
! {bedtoolsDirectory}intersectBed \
-u \
-a {exonList} \
-b {CGMotifList} \
| wc -l
!echo "CG motif overlaps with exons"

  636270
CG motif overlaps with exons


Proportion exon overlap with CG motifs:

In [69]:
636270/14458703

0.04400602184027157

In [71]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a {exonList} \
-b {CGMotifList} \
> 2018-11-07-Exon-CGmotif.txt

In [72]:
!head 2018-11-07-Exon-CGmotif.txt

NC_035780.1	13597	13599	NC_035780.1	13597	13599	CG_motif
NC_035780.1	28992	28994	NC_035780.1	28992	28994	CG_motif
NC_035780.1	29001	29003	NC_035780.1	29001	29003	CG_motif
NC_035780.1	29028	29030	NC_035780.1	29028	29030	CG_motif
NC_035780.1	30539	30541	NC_035780.1	30539	30541	CG_motif
NC_035780.1	30574	30576	NC_035780.1	30574	30576	CG_motif
NC_035780.1	30602	30604	NC_035780.1	30602	30604	CG_motif
NC_035780.1	30676	30678	NC_035780.1	30676	30678	CG_motif
NC_035780.1	30695	30697	NC_035780.1	30695	30697	CG_motif
NC_035780.1	30723	30725	NC_035780.1	30723	30725	CG_motif


### 3b. Introns

In [73]:
! {bedtoolsDirectory}intersectBed \
-u \
-a {intronList} \
-b {CGMotifList} \
| wc -l
!echo "CG motif overlaps with introns"

  245500
CG motif overlaps with introns


Proportion intron overlap with CG motifs:

In [74]:
245500/14458703

0.016979392964915317

In [75]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a {intronList} \
-b {CGMotifList} \
> 2018-11-07-Intron-CGmotif.txt

In [76]:
!head 2018-11-07-Intron-CGmotif.txt

NC_035780.1	29180	29182	NC_035780.1	29180	29182	CG_motif
NC_035780.1	29203	29205	NC_035780.1	29203	29205	CG_motif
NC_035780.1	29221	29223	NC_035780.1	29221	29223	CG_motif
NC_035780.1	29295	29297	NC_035780.1	29295	29297	CG_motif
NC_035780.1	29323	29325	NC_035780.1	29323	29325	CG_motif
NC_035780.1	29326	29328	NC_035780.1	29326	29328	CG_motif
NC_035780.1	29412	29414	NC_035780.1	29412	29414	CG_motif
NC_035780.1	29452	29454	NC_035780.1	29452	29454	CG_motif
NC_035780.1	29672	29674	NC_035780.1	29672	29674	CG_motif
NC_035780.1	29758	29760	NC_035780.1	29758	29760	CG_motif


### 3c. mRNA

In [77]:
! {bedtoolsDirectory}intersectBed \
-u \
-a {mRNAList} \
-b {CGMotifList} \
| wc -l
!echo "CG motif overlaps with mRNA"

   60195
CG motif overlaps with mRNA


Proportion mRNA overlap with CG motifs:

In [78]:
60195/14458703

0.004163236495002352

In [79]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a {mRNAList} \
-b {CGMotifList} \
> 2018-11-07-mRNA-CGmotif.txt

In [80]:
!head 2018-11-07-mRNA-CGmotif.txt

NC_035780.1	Gnomon	mRNA	28993	28994	.	+	.	ID=rna1;Parent=gene1;Dbxref=GeneID:111126949,Genbank:XM_022471938.1;Name=XM_022471938.1;gbkey=mRNA;gene=LOC111126949;model_evidence=Supporting evidence includes similarity to: 3 Proteins%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 21 samples with support for all annotated introns;product=UNC5C-like protein;transcript_id=XM_022471938.1	NC_035780.1	28992	28994	CG_motif
NC_035780.1	Gnomon	mRNA	29002	29003	.	+	.	ID=rna1;Parent=gene1;Dbxref=GeneID:111126949,Genbank:XM_022471938.1;Name=XM_022471938.1;gbkey=mRNA;gene=LOC111126949;model_evidence=Supporting evidence includes similarity to: 3 Proteins%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 21 samples with support for all annotated introns;product=UNC5C-like protein;transcript_id=XM_022471938.1	NC_035780.1	29001	29003	CG_motif
NC_035780.1	Gnomon	mRNA	29029	29030	.	+	.	ID=rna1;Parent=gene1;Dbxref=GeneID:1111269

### 3d. Transposable Elements (All)

In [92]:
! {bedtoolsDirectory}intersectBed \
-u \
-a {transposableElementsAll} \
-b {CGMotifList} \
| wc -l
!echo "CG motif overlaps with transposable elements (all)"

  438365
CG motif overlaps with transposable elements (all)


Proportion TE (all) overlap with CG motifs:

In [93]:
438365/14458703

0.03031841791065215

In [94]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a {transposableElementsAll} \
-b {CGMotifList} \
> 2018-11-07-TE-all-CGmotif.txt

In [95]:
!head 2018-11-07-TE-all-CGmotif.txt

NC_007175.2	RepeatMasker	similarity	264	265	31.1	+	.	Target "Motif:REP-6_LMi" 2920 4055	NC_007175.2	263	265	CG_motif
NC_007175.2	RepeatMasker	similarity	266	267	31.1	+	.	Target "Motif:REP-6_LMi" 2920 4055	NC_007175.2	265	267	CG_motif
NC_007175.2	RepeatMasker	similarity	332	333	31.1	+	.	Target "Motif:REP-6_LMi" 2920 4055	NC_007175.2	331	333	CG_motif
NC_007175.2	RepeatMasker	similarity	367	368	31.1	+	.	Target "Motif:REP-6_LMi" 2920 4055	NC_007175.2	366	368	CG_motif
NC_007175.2	RepeatMasker	similarity	473	474	31.1	+	.	Target "Motif:REP-6_LMi" 2920 4055	NC_007175.2	472	474	CG_motif
NC_007175.2	RepeatMasker	similarity	592	593	31.1	+	.	Target "Motif:REP-6_LMi" 2920 4055	NC_007175.2	591	593	CG_motif
NC_007175.2	RepeatMasker	similarity	665	666	31.1	+	.	Target "Motif:REP-6_LMi" 2920 4055	NC_007175.2	664	666	CG_motif
NC_007175.2	RepeatMasker	similarity	684	685	31.1	+	.	Target "Motif:REP-6_LMi" 2920 4055	NC_007175.2	683	685	CG_motif
NC_007175.2	RepeatMasker	similarity	709	710	31.1	+	.	Target "Mot

### 3e. Transposable Elements (_C. gigas_ only)

In [23]:
! {bedtoolsDirectory}intersectBed \
-u \
-a {transposableElementsCg} \
-b {CGMotifList} \
| wc -l
!echo "CG motif overlaps with transposable elements (Cg)"

  372479
CG motif overlaps with transposable elements (Cg)


Proportion TE (Cv) overlap with CG motifs:

In [97]:
372479/14458703

0.025761577646349055

In [25]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a {transposableElementsCg} \
-b {CGMotifList} \
> 2018-11-07-TE-Cg-CGmotif.txt

In [26]:
!head 2018-11-07-TE-Cg-CGmotif.txt

NC_007175.2	RepeatMasker	similarity	1874	1875	33.6	+	.	Target "Motif:LSU-rRNA_Cel" 2372 2520	NC_007175.2	1873	1875	CG_motif
NC_007175.2	RepeatMasker	similarity	1919	1920	33.6	+	.	Target "Motif:LSU-rRNA_Cel" 2372 2520	NC_007175.2	1918	1920	CG_motif
NC_007175.2	RepeatMasker	similarity	2004	2005	33.6	+	.	Target "Motif:LSU-rRNA_Cel" 2372 2520	NC_007175.2	2003	2005	CG_motif
NC_035780.1	RepeatMasker	similarity	5080	5080	32.5	-	.	Target "Motif:Gypsy-62_CGi-I" 2102 4631	NC_035780.1	5078	5080	CG_motif
NC_035780.1	RepeatMasker	similarity	5160	5161	32.5	-	.	Target "Motif:Gypsy-62_CGi-I" 2102 4631	NC_035780.1	5159	5161	CG_motif
NC_035780.1	RepeatMasker	similarity	5163	5164	32.5	-	.	Target "Motif:Gypsy-62_CGi-I" 2102 4631	NC_035780.1	5162	5164	CG_motif
NC_035780.1	RepeatMasker	similarity	5175	5176	32.5	-	.	Target "Motif:Gypsy-62_CGi-I" 2102 4631	NC_035780.1	5174	5176	CG_motif
NC_035780.1	RepeatMasker	similarity	5192	5193	32.5	-	.	Target "Motif:Gypsy-62_CGi-I" 2102 4631	NC_035780.1	5191	5193	CG_moti

## 4. Identify Overlaps between Transposable Elements and Other Genome Feature Tracks

To fully understand my results, I also need to know where TEs are located with respect to exons, introns, and mRNA coding regions.

### 4a. Transposable Elements (All)

#### Exons

In [100]:
! {bedtoolsDirectory}intersectBed \
-u \
-a {exonList} \
-b {transposableElementsAll} \
| wc -l
!echo "Exon overlaps with transposable elements (all)"

   50331
Exon overlaps with transposable elements (all)


Proportion exon overlap with TE (all):

In [101]:
50331/692371

0.07269368589961163

In [102]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a {exonList} \
-b {transposableElementsAll} \
> 2018-11-07-Exon-TE-all.txt

In [103]:
!head 2018-11-07-Exon-TE-all.txt

NC_035780.1	109967	109996	NC_035780.1	RepeatMasker	similarity	109968	109996	 0.0	+	.	Target "Motif:(CCT)n" 1 29
NC_035780.1	164885	164914	NC_035780.1	RepeatMasker	similarity	164886	164914	 7.3	+	.	Target "Motif:(GAG)n" 1 29
NC_035780.1	166074	166280	NC_035780.1	RepeatMasker	similarity	166075	166280	32.8	+	.	Target "Motif:Harbinger1_DR" 1472 1676
NC_035780.1	166500	166566	NC_035780.1	RepeatMasker	similarity	166501	166566	30.3	+	.	Target "Motif:Harbinger-6_DR" 1152 1217
NC_035780.1	166597	166642	NC_035780.1	RepeatMasker	similarity	166598	166642	17.8	+	.	Target "Motif:hATw-1_HM" 2778 2822
NC_035780.1	220121	220199	NC_035780.1	RepeatMasker	similarity	220122	220199	24.7	-	.	Target "Motif:Gypsy-75_CQ-I" 1012 1091
NC_035780.1	228341	228392	NC_035780.1	RepeatMasker	similarity	228342	228392	20.0	+	.	Target "Motif:RTE-3_Hmel" 1405 1455
NC_035780.1	227767	227819	NC_035780.1	RepeatMasker	similarity	227768	227819	25.0	+	.	Target "Motif:A-rich" 1 54
NC_035780.1	228341	228392	NC_035780.1	RepeatMasker

#### Introns

In [105]:
!{bedtoolsDirectory}intersectBed \
-u \
-a {intronList} \
-b {transposableElementsAll} \
| wc -l
!echo "Intron overlaps with transposable elements (all)"

  105643
Intron overlaps with transposable elements (all)


Proportion intron overlap with TE (all):

In [106]:
105643/692371

0.1525814917147021

In [107]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a {intronList} \
-b {transposableElementsAll} \
> 2018-11-07-Intron-TE-all.txt

In [108]:
!head 2018-11-07-Intron-TE-all.txt

NC_035780.1	32719	32819	NC_035780.1	RepeatMasker	similarity	32720	32819	18.2	+	.	Target "Motif:Crypton-9N1_CGi" 239 337
NC_035780.1	48462	48520	NC_035780.1	RepeatMasker	similarity	48463	48520	 8.8	+	.	Target "Motif:BivaMD-SINE1_CrVi" 280 337
NC_035780.1	48665	49000	NC_035780.1	RepeatMasker	similarity	48666	49000	10.9	-	.	Target "Motif:BivaMD-SINE1_CrVi" 1 337
NC_035780.1	50250	50279	NC_035780.1	RepeatMasker	similarity	50251	50279	 0.0	+	.	Target "Motif:(GGTTAG)n" 1 29
NC_035780.1	50605	50760	NC_035780.1	RepeatMasker	similarity	50606	50760	21.3	+	.	Target "Motif:Harbinger-2N1_CGi" 1 166
NC_035780.1	50976	51034	NC_035780.1	RepeatMasker	similarity	50977	51034	 0.0	+	.	Target "Motif:(TA)n" 1 58
NC_035780.1	51455	51498	NC_035780.1	RepeatMasker	similarity	51456	51498	 0.0	+	.	Target "Motif:(AG)n" 1 43
NC_035780.1	51720	51922	NC_035780.1	RepeatMasker	similarity	51721	51922	21.8	+	.	Target "Motif:Harbinger-2N1_CGi" 2568 2776
NC_035780.1	53155	53294	NC_035780.1	RepeatMasker	similarity	53156	532

#### mRNA

In [109]:
!{bedtoolsDirectory}intersectBed \
-u \
-a {mRNAList} \
-b {transposableElementsAll} \
| wc -l
!echo "mRNA overlaps with transposable elements (all)"

   55069
mRNA overlaps with transposable elements (all)


Proportion mRNA overlap with TE (all):

In [110]:
55069/692371

0.07953683790915564

In [111]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a {mRNAList} \
-b {transposableElementsAll} \
> 2018-11-07-mRNA-TE-all.txt

In [112]:
!head 2018-11-07-mRNA-TE-all.txt

NC_035780.1	Gnomon	mRNA	32720	32819	.	+	.	ID=rna1;Parent=gene1;Dbxref=GeneID:111126949,Genbank:XM_022471938.1;Name=XM_022471938.1;gbkey=mRNA;gene=LOC111126949;model_evidence=Supporting evidence includes similarity to: 3 Proteins%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 21 samples with support for all annotated introns;product=UNC5C-like protein;transcript_id=XM_022471938.1	NC_035780.1	RepeatMasker	similarity	32720	32819	18.2	+	.	Target "Motif:Crypton-9N1_CGi" 239 337
NC_035780.1	Gnomon	mRNA	48463	48520	.	-	.	ID=rna2;Parent=gene2;Dbxref=GeneID:111110729,Genbank:XM_022447324.1;Name=XM_022447324.1;gbkey=mRNA;gene=LOC111110729;model_evidence=Supporting evidence includes similarity to: 1 Protein%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments;product=FMRFamide receptor-like%2C transcript variant X1;transcript_id=XM_022447324.1	NC_035780.1	RepeatMasker	similarity	48463	48520	 8.8	+	.	Target "Motif:BivaMD-SINE1_CrVi

### 4b. Transposable Elements (_C. virginica_ only)

#### Exons

In [24]:
! {bedtoolsDirectory}intersectBed \
-u \
-a {exonList} \
-b {transposableElementsCg} \
| wc -l
!echo "Exon overlaps with transposable elements (Cg)"

   41511
Exon overlaps with transposable elements (Cg)


Proportion exon overlap with TE (Cg):

In [115]:
41511/692371

0.059954850795310606

In [27]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a {exonList} \
-b {transposableElementsCg} \
> 2018-11-07-Exon-TE-Cg.txt

In [28]:
!head 2018-11-07-Exon-TE-Cg.txt

NC_035780.1	109967	109996	NC_035780.1	RepeatMasker	similarity	109968	109996	 0.0	+	.	Target "Motif:(CCT)n" 1 29
NC_035780.1	164885	164914	NC_035780.1	RepeatMasker	similarity	164886	164914	 7.3	+	.	Target "Motif:(GAG)n" 1 29
NC_035780.1	227767	227819	NC_035780.1	RepeatMasker	similarity	227768	227819	25.0	+	.	Target "Motif:A-rich" 1 54
NC_035780.1	227767	227819	NC_035780.1	RepeatMasker	similarity	227768	227819	25.0	+	.	Target "Motif:A-rich" 1 54
NC_035780.1	227767	227819	NC_035780.1	RepeatMasker	similarity	227768	227819	25.0	+	.	Target "Motif:A-rich" 1 54
NC_035780.1	233475	233478	NC_035780.1	RepeatMasker	similarity	233445	233478	10.1	+	.	Target "Motif:(CCTTT)n" 1 35
NC_035780.1	232863	233028	NC_035780.1	RepeatMasker	similarity	232798	233028	29.7	-	.	Target "Motif:ISL2EU-N8_CGi" 15 237
NC_035780.1	269562	269603	NC_035780.1	RepeatMasker	similarity	269563	269603	17.1	+	.	Target "Motif:(ATG)n" 1 42
NC_035780.1	258539	258574	NC_035780.1	RepeatMasker	similarity	258540	258574	16.3	+	.	Target "

#### Introns

In [29]:
! {bedtoolsDirectory}intersectBed \
-u \
-a {intronList} \
-b {transposableElementsCg} \
| wc -l
!echo "Intron overlaps with transposable elements (Cg)"

   98494
Intron overlaps with transposable elements (Cg)


Proportion intron overlap with TE (Cg):

In [119]:
98494/692371

0.1422561025808418

In [30]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a {intronList} \
-b {transposableElementsCg} \
> 2018-11-07-Intron-TE-Cg.txt

In [31]:
!head 2018-11-07-Intron-TE-Cg.txt

NC_035780.1	32719	32819	NC_035780.1	RepeatMasker	similarity	32720	32819	18.2	+	.	Target "Motif:Crypton-9N1_CGi" 239 337
NC_035780.1	46753	46805	NC_035780.1	RepeatMasker	similarity	46754	46805	 6.8	+	.	Target "Motif:DNA-22_CGi" 631 722
NC_035780.1	50250	50279	NC_035780.1	RepeatMasker	similarity	50251	50279	 0.0	+	.	Target "Motif:(GGTTAG)n" 1 29
NC_035780.1	50605	50760	NC_035780.1	RepeatMasker	similarity	50606	50760	21.3	+	.	Target "Motif:Harbinger-2N1_CGi" 1 166
NC_035780.1	50976	51034	NC_035780.1	RepeatMasker	similarity	50977	51034	 0.0	+	.	Target "Motif:(TA)n" 1 58
NC_035780.1	51455	51498	NC_035780.1	RepeatMasker	similarity	51456	51498	 0.0	+	.	Target "Motif:(AG)n" 1 43
NC_035780.1	51720	51922	NC_035780.1	RepeatMasker	similarity	51721	51922	21.8	+	.	Target "Motif:Harbinger-2N1_CGi" 2568 2776
NC_035780.1	86839	86942	NC_035780.1	RepeatMasker	similarity	86840	86942	27.4	-	.	Target "Motif:Helitron-N14_CGi" 83 189
NC_035780.1	87408	87513	NC_035780.1	RepeatMasker	similarity	87409	87513	19.8

#### mRNA

In [32]:
! {bedtoolsDirectory}intersectBed \
-u \
-a {mRNAList} \
-b {transposableElementsCg} \
| wc -l
!echo "mRNA overlaps with transposable elements (Cg)"

   53914
mRNA overlaps with transposable elements (Cg)


Proportion mRNA overlap with TE (Cv):

In [123]:
53914/692371

0.07786865712168765

In [33]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a {mRNAList} \
-b {transposableElementsCg} \
> 2018-11-07-mRNA-TE-Cg.txt

In [34]:
!head 2018-11-07-Intron-TE-Cg.txt

NC_035780.1	32719	32819	NC_035780.1	RepeatMasker	similarity	32720	32819	18.2	+	.	Target "Motif:Crypton-9N1_CGi" 239 337
NC_035780.1	46753	46805	NC_035780.1	RepeatMasker	similarity	46754	46805	 6.8	+	.	Target "Motif:DNA-22_CGi" 631 722
NC_035780.1	50250	50279	NC_035780.1	RepeatMasker	similarity	50251	50279	 0.0	+	.	Target "Motif:(GGTTAG)n" 1 29
NC_035780.1	50605	50760	NC_035780.1	RepeatMasker	similarity	50606	50760	21.3	+	.	Target "Motif:Harbinger-2N1_CGi" 1 166
NC_035780.1	50976	51034	NC_035780.1	RepeatMasker	similarity	50977	51034	 0.0	+	.	Target "Motif:(TA)n" 1 58
NC_035780.1	51455	51498	NC_035780.1	RepeatMasker	similarity	51456	51498	 0.0	+	.	Target "Motif:(AG)n" 1 43
NC_035780.1	51720	51922	NC_035780.1	RepeatMasker	similarity	51721	51922	21.8	+	.	Target "Motif:Harbinger-2N1_CGi" 2568 2776
NC_035780.1	86839	86942	NC_035780.1	RepeatMasker	similarity	86840	86942	27.4	-	.	Target "Motif:Helitron-N14_CGi" 83 189
NC_035780.1	87408	87513	NC_035780.1	RepeatMasker	similarity	87409	87513	19.8

## 5. Calculate Overlap Proportions

It's important to understand how many overlaps are present between various feature tracks and CG motifs. CG motifs are where we expect methylation to happen. If there are more overlaps present bewteen a certain feature and the CG motifs, we would expect to see most of our DMLs in that region. I also want to understand overlap proportions with DMLS. 

Here are the questions I will answer:

1. Out the total number of CG motifs, how many overlaped with a feature track?
2. Out of the total number of transposable elements, how many overlaped with a feature track?
2. What proportion of total overlaps does a certain feature track represent?
3. Out the total number of DML, how many overlaped with a feature track?
5. Out of the total number of DMR, how many overlaped with a feature track?

### 5a. CG motif Overlaps with Feature Tracks

I already calculated the numbers associated with the first question in the first section. I'll remind you of those numbers:

- Proportion exon overlap with CG motifs: 4.40% (0.04400602184027157)
- Proportion intron overlap with CG motifs: 1.70% (0.016979392964915317)
- Proportion mRNA overlap with CG motifs: 0.42% (0.004163236495002352)
- Proportion transposable element (all) overlap with CG motifs: 2.58% (0.025761577646349055) 
- Proportion transposable element (_C. gigas_) overlap with CG motifs: 3.03% (0.03031841791065215)

### 5b. Transposable Element (all) Overlaps with Feature Tracks

See 4a for more details.

- Proportion exon overlap with TE (all): 7.27% (0.07269368589961163)
- Proportion intron overlap with TE (all): 15.3% (0.1525814917147021)
- Proportion mRNA overlap with TE (all): 7.95% (0.07953683790915564)

### 5c. Transposable Element (_C. gigas_ only) Overlaps with Feature Tracks

See 4b for more details.

- Proportion exon overlap with TE (Cg): 6.00% (0.059954850795310606)
- Proportion intron overlap with TE (Cg): 14.2% (0.1422561025808418)
- Proportion mRNA overlap with TE (Cg): 7.79% (0.07786865712168765)

### 5d. Proportion Total Overlaps by Feature Track

Since I have two different transposable element tracks, I'll repeat these calculations for each track.

#### Transposable Elements (all)

First, I need to calculate the total number of overlaps we had:

(exon overlap with CG motifs) + (intron overlap with CG motifs) + (mRNA overlap with CG motifs) + (TE overlap with CG motifs)

In [126]:
636270 + 245500 + 60195 + 438365

1380330

Now, I calculate the proportions:

Exons:

In [127]:
636270/1380330

0.4609549890243637

Introns:

In [128]:
245500/1380330

0.17785601993726138

mRNA:

In [129]:
60195/1380330

0.04360913694551303

Transposable Elements (all):

In [130]:
438365/1380330

0.3175798540928619

- Proportion exon overlap out of total overlaps: 46.10% (0.4609549890243637)
- Proportion intron overlap out of total overlaps: 17.79% (0.17785601993726138)
- Proportion mRNA overlap out of total overlaps: 4.36% (0.04360913694551303)
- Proportion transposable elements (all) overlap out of total overlaps: 31.76% (0.3175798540928619)

#### Transposable Elements (_C. gigas_ only)

Total number of overlaps:

In [131]:
636270 + 245500 + 60195 + 372479

1314444

Exons:

In [132]:
636270/1314444

0.4840601805782521

Introns:

In [133]:
245500/1314444

0.18677098453794913

mRNA:

In [134]:
60195/1314444

0.04579502816399938

Tranposable Elements (_C. gigas_ only):

In [135]:
372479/1314444

0.2833738067197994

- Proportion exon overlap out of total overlaps: 48.41% (0.4840601805782521)
- Proportion intron overlap out of total overlaps: 18.68% (0.18677098453794913)
- Proportion mRNA overlap out of total overlaps: 4.58% (0.04579502816399938)
- Proportion transposable elements (Cg) overlap out of total overlaps: 28.34% (0.2833738067197994)

### 5e. DML Overlaps with Feature Tracks

Exons:

In [44]:
388/630

0.6158730158730159

Introns:

In [45]:
200/630

0.31746031746031744

mRNA:

In [46]:
576/630

0.9142857142857143

Transposable Elements (all):

In [47]:
61/630

0.09682539682539683

Transposable elements (_C. gigas_ only):

In [48]:
40/630

0.06349206349206349

- Proportion exon overlap with DMLs: 61.59% (0.6158730158730159)
- Proportion intron overlap with DMLs: 31.75% (0.31746031746031744)
- Proportion mRNA overlap with DMLs: 91.43% (0.9142857142857143)
- Proportion transposable element (all) overlap with DMLs: 9.68% (0.09682539682539683)
- Proportion transposable element (_C. gigas_) overlap with DMLs: 6.35% (0.06349206349206349)

### 5f. DMR Overlaps with Feature Tracks

Exons:

In [19]:
64/162

0.3950617283950617

Introns:

In [20]:
112/162

0.691358024691358

mRNA:

In [21]:
139/162

0.8580246913580247

Transposable Elements (All):

In [142]:
39/162

0.24074074074074073

Transposable Elements (_C. gigas_ only)

In [143]:
23/162

0.1419753086419753

- Proportion exon overlap with DMRs: 39.51% (0.3950617283950617)
- Proportion intron overlap with DMRs: 69.14% (0.691358024691358)
- Proportion mRNA overlap with DMRs: 85.80% (0.8580246913580247)
- Proportion transposable element (all) overlap with DMRs: 24.07% (0.24074074074074073)
- Proportion transposable element (_C. gigas_ only) overlap with DMRs: 14.20% (0.1419753086419753)

## 6. Gene Flanking

I will perform a flanking analysis in two ways. First, I will use `bedtools flank` to add 1000 bp regions to each mRNA coding region. I can then isolate these flanks and intersect them with various genomic feature files. Second I will use `bedtools closest` to find the closest non-overlapping DML or DMR to each mRNA coding region.

In [36]:
mkdir 2018-11-14-Flanking-Analysis #Create a new directory for flanking analysis output

In [37]:
ls -F

2018-11-07-DML-Exon.txt
2018-11-07-DML-Intron.txt
2018-11-07-DML-TE-Cg.txt
2018-11-07-DML-TE-all.txt
2018-11-07-DML-mRNA.txt
2018-11-07-DMR-Exon.txt
2018-11-07-DMR-Intron.txt
2018-11-07-DMR-TE-Cg.txt
2018-11-07-DMR-TE-all.txt
2018-11-07-DMR-mRNA.txt
2018-11-07-Exon-CGmotif.txt
2018-11-07-Exon-TE-Cg.txt
2018-11-07-Exon-TE-all.txt
2018-11-07-Intron-CGmotif.txt
2018-11-07-Intron-TE-Cg.txt
2018-11-07-Intron-TE-all.txt
2018-11-07-TE-Cg-CGmotif.txt
2018-11-07-TE-all-CGmotif.txt
2018-11-07-Unique-Genes-in-DML-mRNA-Overlap.txt
2018-11-07-Unique-Genes-in-DMR-mRNA-Overlap.txt
2018-11-07-mRNA-CGmotif.txt
2018-11-07-mRNA-TE-Cg.txt
2018-11-07-mRNA-TE-all.txt
2018-11-14-Flanking-Analysis/
C_virginica-3.0_CG-motif.bed
C_virginica-3.0_Gnomon_exon.bed
C_virginica-3.0_Gnomon_mRNA.gff3
C_virginica-3.0_TE-Cg.gff
C_virginica-3.0_TE-all.gff
C_virginica-3.0_intron.bed


### 6a. `flank`

I also need to know if DMLs and CG motifs overlap with regions that flank mRNA. These flanking regions could be promoters or transcription factors that could regulate these processes. To do this, I will use `bedtools flank`:

1. Path to `flankBed`
2. -i: Path to mRNA GFF file
3. -g: Path to C. virginica "genome" file. flankBed requires the start and stop position of each genome (see this issue). I created a file like in TextWrangler using chromosome lengths from NCBI.
4. -b 1000: Add 1000 bp flanks to each end of the coding region

In [47]:
! {bedtoolsDirectory}flankBed \
-i {mRNAList} \
-g 2018-11-14-Flanking-Analysis/2018-11-14-bedtools-Chromosome-Length.txt \
-b 1000 \
> 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-1000bp-Flanks.bed

In [98]:
!head {mRNAList} #The original file, just for comparison

NC_035780.1	Gnomon	mRNA	28961	33324	.	+	.	ID=rna1;Parent=gene1;Dbxref=GeneID:111126949,Genbank:XM_022471938.1;Name=XM_022471938.1;gbkey=mRNA;gene=LOC111126949;model_evidence=Supporting evidence includes similarity to: 3 Proteins%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 21 samples with support for all annotated introns;product=UNC5C-like protein;transcript_id=XM_022471938.1
NC_035780.1	Gnomon	mRNA	43111	66897	.	-	.	ID=rna2;Parent=gene2;Dbxref=GeneID:111110729,Genbank:XM_022447324.1;Name=XM_022447324.1;gbkey=mRNA;gene=LOC111110729;model_evidence=Supporting evidence includes similarity to: 1 Protein%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments;product=FMRFamide receptor-like%2C transcript variant X1;transcript_id=XM_022447324.1
NC_035780.1	Gnomon	mRNA	43111	46506	.	-	.	ID=rna3;Parent=gene2;Dbxref=GeneID:111110729,Genbank:XM_022447333.1;Name=XM_022447333.1;gbkey=mRNA;gene=LOC111110729;model_evidence=Supporting

In [99]:
!head 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-1000bp-Flanks.bed #Isolated flanks. The first entry is the upstream flank for the first mRNA coding region, second is the downstream flank for the mRNA coding region, etc.

NC_035780.1	Gnomon	mRNA	27961	28960	.	+	.	ID=rna1;Parent=gene1;Dbxref=GeneID:111126949,Genbank:XM_022471938.1;Name=XM_022471938.1;gbkey=mRNA;gene=LOC111126949;model_evidence=Supporting evidence includes similarity to: 3 Proteins%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 21 samples with support for all annotated introns;product=UNC5C-like protein;transcript_id=XM_022471938.1
NC_035780.1	Gnomon	mRNA	33325	34324	.	+	.	ID=rna1;Parent=gene1;Dbxref=GeneID:111126949,Genbank:XM_022471938.1;Name=XM_022471938.1;gbkey=mRNA;gene=LOC111126949;model_evidence=Supporting evidence includes similarity to: 3 Proteins%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 21 samples with support for all annotated introns;product=UNC5C-like protein;transcript_id=XM_022471938.1
NC_035780.1	Gnomon	mRNA	42111	43110	.	-	.	ID=rna2;Parent=gene2;Dbxref=GeneID:111110729,Genbank:XM_022447324.1;Name=XM_022447324.1;gbkey=mRNA;gene=LOC1

In [23]:
!wc -l 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-1000bp-Flanks.bed

  120400 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-1000bp-Flanks.bed


Now that I have these flanks, I want to separate the upstream flank from the downstream flank. I will do this using `awk`. If th row number is odd, the rows go into the upstream flank file. If the row number is even, it goes into the downstream flank file.

In [26]:
!awk '{ if (NR%2) print > "2018-11-14-Flanking-Analysis/2018-11-15-mRNA-Upstream-Flanks.bed"; \
else print > "2018-11-14-Flanking-Analysis/2018-11-15-mRNA-Downstream-Flanks.bed" }' \
2018-11-14-Flanking-Analysis/2018-11-14-mRNA-1000bp-Flanks.bed

#### Upstream flank

In [28]:
!head 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-Upstream-Flanks.bed

NC_035780.1	Gnomon	mRNA	27961	28960	.	+	.	ID=rna1;Parent=gene1;Dbxref=GeneID:111126949,Genbank:XM_022471938.1;Name=XM_022471938.1;gbkey=mRNA;gene=LOC111126949;model_evidence=Supporting evidence includes similarity to: 3 Proteins%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 21 samples with support for all annotated introns;product=UNC5C-like protein;transcript_id=XM_022471938.1
NC_035780.1	Gnomon	mRNA	42111	43110	.	-	.	ID=rna2;Parent=gene2;Dbxref=GeneID:111110729,Genbank:XM_022447324.1;Name=XM_022447324.1;gbkey=mRNA;gene=LOC111110729;model_evidence=Supporting evidence includes similarity to: 1 Protein%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments;product=FMRFamide receptor-like%2C transcript variant X1;transcript_id=XM_022447324.1
NC_035780.1	Gnomon	mRNA	42111	43110	.	-	.	ID=rna3;Parent=gene2;Dbxref=GeneID:111110729,Genbank:XM_022447333.1;Name=XM_022447333.1;gbkey=mRNA;gene=LOC111110729;model_evidence=Supporting

In [29]:
!wc -l 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-Upstream-Flanks.bed

   60200 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-Upstream-Flanks.bed


#### Downstream flanks

In [30]:
!head 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-Downstream-Flanks.bed

NC_035780.1	Gnomon	mRNA	33325	34324	.	+	.	ID=rna1;Parent=gene1;Dbxref=GeneID:111126949,Genbank:XM_022471938.1;Name=XM_022471938.1;gbkey=mRNA;gene=LOC111126949;model_evidence=Supporting evidence includes similarity to: 3 Proteins%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 21 samples with support for all annotated introns;product=UNC5C-like protein;transcript_id=XM_022471938.1
NC_035780.1	Gnomon	mRNA	66898	67897	.	-	.	ID=rna2;Parent=gene2;Dbxref=GeneID:111110729,Genbank:XM_022447324.1;Name=XM_022447324.1;gbkey=mRNA;gene=LOC111110729;model_evidence=Supporting evidence includes similarity to: 1 Protein%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments;product=FMRFamide receptor-like%2C transcript variant X1;transcript_id=XM_022447324.1
NC_035780.1	Gnomon	mRNA	46507	47506	.	-	.	ID=rna3;Parent=gene2;Dbxref=GeneID:111110729,Genbank:XM_022447333.1;Name=XM_022447333.1;gbkey=mRNA;gene=LOC111110729;model_evidence=Supporting

In [31]:
!wc -l 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-Downstream-Flanks.bed

   60200 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-Downstream-Flanks.bed


Now I'll take the upstream and downstream flank BEDfiles I made and use it in intersectBed to find overlaps with DML, DMR, and CG motifs!

1. Path to intersectBed
2. -wb: Write output according to the second file
3. -a: Path to BEDfile created with flanks
4. -b: Specify either DML, DMR, or CG motif file. Overlaps between the flanks and CG motifs can be used as a background when comparing DML-flank and DMR-flank results
5. ">" filename: Redirect output to a .txt file

#### DML

In [32]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-Upstream-Flanks.bed \
-b {DMLlist} \
> 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-100bp-UpstreamFlanks-DML.txt

In [33]:
!head 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-100bp-UpstreamFlanks-DML.txt

NC_035780.1	Gnomon	mRNA	8833125	8833126	.	+	.	ID=rna875;Parent=gene522;Dbxref=GeneID:111138488,Genbank:XM_022490485.1;Name=XM_022490485.1;gbkey=mRNA;gene=LOC111138488;model_evidence=Supporting evidence includes similarity to: 5 Proteins%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 26 samples with support for all annotated introns;product=hsp70-binding protein 1-like;transcript_id=XM_022490485.1	NC_035780.1	8833124	8833126	60
NC_035780.1	Gnomon	mRNA	55484706	55484707	.	-	.	ID=rna5598;Parent=gene3286;Dbxref=GeneID:111102644,Genbank:XM_022435478.1;Name=XM_022435478.1;gbkey=mRNA;gene=LOC111102644;model_evidence=Supporting evidence includes similarity to: 3 Proteins%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 14 samples with support for all annotated introns;product=histone-lysine N-methyltransferase KMT5B-like%2C transcript variant X2;transcript_id=XM_022435478.1	NC_035780.1	55484705	55484707	-52
NC_

In [34]:
!wc -l 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-100bp-UpstreamFlanks-DML.txt

      67 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-100bp-UpstreamFlanks-DML.txt


In [35]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-Downstream-Flanks.bed \
-b {DMLlist} \
> 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-100bp-DownstreamFlanks-DML.txt

In [36]:
!head 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-100bp-DownstreamFlanks-DML.txt

NC_035780.1	Gnomon	mRNA	1882692	1882693	.	+	.	ID=rna154;Parent=gene94;Dbxref=GeneID:111102439,Genbank:XM_022435187.1;Name=XM_022435187.1;gbkey=mRNA;gene=LOC111102439;model_evidence=Supporting evidence includes similarity to: 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 5 samples with support for all annotated introns;product=uncharacterized LOC111102439;transcript_id=XM_022435187.1	NC_035780.1	1882691	1882693	64
NC_035781.1	Gnomon	mRNA	20126030	20126031	.	-	.	ID=rna9080;Parent=gene5340;Dbxref=GeneID:111122222,Genbank:XM_022463862.1;Name=XM_022463862.1;gbkey=mRNA;gene=LOC111122222;model_evidence=Supporting evidence includes similarity to: 1 Protein%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 18 samples with support for all annotated introns;product=phosphatidylinositol N-acetylglucosaminyltransferase subunit Q-like;transcript_id=XM_022463862.1	NC_035781.1	20126029	20126031	-52
NC_035781.1	Gnomon	mRNA	289

In [37]:
!wc -l 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-100bp-DownstreamFlanks-DML.txt

      49 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-100bp-DownstreamFlanks-DML.txt


#### DMR

In [38]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-Upstream-Flanks.bed \
-b {DMRlist} \
> 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-100bp-UpstreamFlanks-DMR.txt

In [39]:
!head 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-100bp-UpstreamFlanks-DMR.txt

NC_035780.1	Gnomon	mRNA	46044701	46044800	.	-	.	ID=rna4633;Parent=gene2726;Dbxref=GeneID:111110504,Genbank:XM_022447031.1;Name=XM_022447031.1;gbkey=mRNA;gene=LOC111110504;model_evidence=Supporting evidence includes similarity to: 1 Protein%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 1 sample with support for all annotated introns;product=uncharacterized LOC111110504%2C transcript variant X1;transcript_id=XM_022447031.1	NC_035780.1	46044700	46044800	DMR	-53
NC_035780.1	Gnomon	mRNA	46044701	46044800	.	-	.	ID=rna4634;Parent=gene2726;Dbxref=GeneID:111110504,Genbank:XM_022447040.1;Name=XM_022447040.1;gbkey=mRNA;gene=LOC111110504;model_evidence=Supporting evidence includes similarity to: 1 Protein%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 1 sample with support for all annotated introns;product=uncharacterized LOC111110504%2C transcript variant X2;transcript_id=XM_022447040.1	NC_035780.1	46044700	460

In [40]:
!wc -l 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-100bp-UpstreamFlanks-DMR.txt

      12 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-100bp-UpstreamFlanks-DMR.txt


In [41]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-Downstream-Flanks.bed \
-b {DMRlist} \
> 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-100bp-DownstreamFlanks-DMR.txt

In [42]:
!head 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-100bp-DownstreamFlanks-DMR.txt

NC_035780.1	Gnomon	mRNA	573701	573800	.	+	.	ID=rna48;Parent=gene35;Dbxref=GeneID:111114201,Genbank:XM_022452489.1;Name=XM_022452489.1;Note=The sequence of the model RefSeq transcript was modified relative to this genomic sequence to represent the inferred CDS: inserted 2 bases in 2 codons;exception=unclassified transcription discrepancy;gbkey=mRNA;gene=LOC111114201;model_evidence=Supporting evidence includes similarity to: 4 Proteins%2C and 99%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 9 samples with support for all annotated introns;product=vacuolar protein sorting-associated protein 13B-like;transcript_id=XM_022452489.1	NC_035780.1	573700	573800	DMR	52
NC_035780.1	Gnomon	mRNA	33945901	33946000	.	-	.	ID=rna3547;Parent=gene2078;Dbxref=GeneID:111131167,Genbank:XM_022478582.1;Name=XM_022478582.1;gbkey=mRNA;gene=LOC111131167;model_evidence=Supporting evidence includes similarity to: 5 Proteins%2C and 100%25 coverage of the annotated genomic feature by R

In [43]:
!wc -l 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-100bp-DownstreamFlanks-DMR.txt

      18 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-100bp-DownstreamFlanks-DMR.txt


#### CG motifs

In [44]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-Upstream-Flanks.bed \
-b {CGMotifList} \
> 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-100bp-UpstreamFlanks-CGmotif.txt

In [45]:
!head 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-100bp-UpstreamFlanks-CGmotif.txt

NC_035780.1	Gnomon	mRNA	27970	27971	.	+	.	ID=rna1;Parent=gene1;Dbxref=GeneID:111126949,Genbank:XM_022471938.1;Name=XM_022471938.1;gbkey=mRNA;gene=LOC111126949;model_evidence=Supporting evidence includes similarity to: 3 Proteins%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 21 samples with support for all annotated introns;product=UNC5C-like protein;transcript_id=XM_022471938.1	NC_035780.1	27969	27971	CG_motif
NC_035780.1	Gnomon	mRNA	27980	27981	.	+	.	ID=rna1;Parent=gene1;Dbxref=GeneID:111126949,Genbank:XM_022471938.1;Name=XM_022471938.1;gbkey=mRNA;gene=LOC111126949;model_evidence=Supporting evidence includes similarity to: 3 Proteins%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 21 samples with support for all annotated introns;product=UNC5C-like protein;transcript_id=XM_022471938.1	NC_035780.1	27979	27981	CG_motif
NC_035780.1	Gnomon	mRNA	28082	28083	.	+	.	ID=rna1;Parent=gene1;Dbxref=GeneID:1111269

In [46]:
!wc -l 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-100bp-UpstreamFlanks-CGmotif.txt

 1286944 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-100bp-UpstreamFlanks-CGmotif.txt


In [47]:
! {bedtoolsDirectory}intersectBed \
-wb \
-a 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-Downstream-Flanks.bed \
-b {CGMotifList} \
> 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-100bp-DownstreamFlanks-CGmotif.txt

In [48]:
!head 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-100bp-DownstreamFlanks-CGmotif.txt

NC_035780.1	Gnomon	mRNA	33408	33409	.	+	.	ID=rna1;Parent=gene1;Dbxref=GeneID:111126949,Genbank:XM_022471938.1;Name=XM_022471938.1;gbkey=mRNA;gene=LOC111126949;model_evidence=Supporting evidence includes similarity to: 3 Proteins%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 21 samples with support for all annotated introns;product=UNC5C-like protein;transcript_id=XM_022471938.1	NC_035780.1	33407	33409	CG_motif
NC_035780.1	Gnomon	mRNA	33452	33453	.	+	.	ID=rna1;Parent=gene1;Dbxref=GeneID:111126949,Genbank:XM_022471938.1;Name=XM_022471938.1;gbkey=mRNA;gene=LOC111126949;model_evidence=Supporting evidence includes similarity to: 3 Proteins%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 21 samples with support for all annotated introns;product=UNC5C-like protein;transcript_id=XM_022471938.1	NC_035780.1	33451	33453	CG_motif
NC_035780.1	Gnomon	mRNA	33481	33482	.	+	.	ID=rna1;Parent=gene1;Dbxref=GeneID:1111269

In [49]:
!wc -l 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-100bp-DownstreamFlanks-CGmotif.txt

 1285210 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-100bp-DownstreamFlanks-CGmotif.txt


### 6b. No overlaps

I also want to count the number of DML or DMR that do not overlap with any features (i.e. DML and DMR in unannotated intergenic regions). To do this, I'll use the `-v` argument in `bedtools`, which reports "those entries in A that have no overlap in B." I can specify multiple files with `-b`. I'll use exons, introns, transposable elements identified using all species, and putative promoter regions (upstream flanks).

#### DML

In [13]:
! {bedtoolsDirectory}intersectBed \
-v \
-a {DMLlist} \
-b {exonList} {intronList} {transposableElementsAll} 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-Upstream-Flanks.bed  \
| wc -l
!echo "DML do not overlap with exons, introns, transposable elements (all), or putative promoters"

      20
DML do not overlap with exons, introns, transposable elements (all), or putative promoters


In [14]:
! {bedtoolsDirectory}intersectBed \
-v \
-a {DMLlist} \
-b {exonList} {intronList} {transposableElementsAll} 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-Upstream-Flanks.bed \
> 2019-03-17-No-Overlap-DML.txt

In [15]:
!head 2019-03-17-No-Overlap-DML.txt

NC_035780.1	55852212	55852214	58
NC_035780.1	60587461	60587463	-74
NC_035781.1	20620123	20620125	57
NC_035781.1	24474567	24474569	53
NC_035781.1	30062222	30062224	60
NC_035781.1	39583208	39583210	-50
NC_035781.1	50711254	50711256	-71
NC_035782.1	58675230	58675232	52
NC_035782.1	65377028	65377030	51
NC_035784.1	2011997	2011999	-60


#### CG motifs

In [10]:
! {bedtoolsDirectory}intersectBed \
-v \
-a {CGMotifList} \
-b {exonList} {intronList} {transposableElementsAll} 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-Upstream-Flanks.bed  \
| wc -l
!echo "CG motifs do not overlap with exons, introns, transposable elements (all), or putative promoters"

 4760788
CG motifs do not overlap with exons, introns, transposable elements (all), or putative promoters


In [11]:
! {bedtoolsDirectory}intersectBed \
-v \
-a {CGMotifList} \
-b {exonList} {intronList} {transposableElementsAll} 2018-11-14-Flanking-Analysis/2018-11-15-mRNA-Upstream-Flanks.bed \
> 2019-03-17-No-Overlap-CGmotifs.txt

In [12]:
!head 2019-03-17-No-Overlap-CGmotifs.txt

NC_035780.1	28	30	CG_motif
NC_035780.1	54	56	CG_motif
NC_035780.1	75	77	CG_motif
NC_035780.1	93	95	CG_motif
NC_035780.1	103	105	CG_motif
NC_035780.1	116	118	CG_motif
NC_035780.1	134	136	CG_motif
NC_035780.1	159	161	CG_motif
NC_035780.1	209	211	CG_motif
NC_035780.1	224	226	CG_motif


### 6c. `closest`

[`bedtools closest`](https://bedtools.readthedocs.io/en/latest/content/tools/closest.html) will find the nearest DML or DMR to an mRNA coding region, but not necessarily a non-overlapping feature. I will use the following code:

1. Path to `closestBed`
2. -io: Ignore features in b that overlap with a
3. -a: Path to mRNA gff
4. -b: Specify either DML, DMR, or CG motif file. The CG motif file will be used as a background for the DML and DMR
6. -t all: In case of a tie, report all matches
7. -D ref: Report distance to A in an extra column. Use negative distances to report upstream features with respect to the reference genome. B features with a lower (start, stop) are upstream.
8. ">" filename: Redirect output to a .txt file

In [46]:
! {bedtoolsDirectory}closestBed \
-io \
-a {mRNAList} \
-b {DMLlist} \
-t all \
-D ref \
> 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-Closest-NoOverlap-DMLs.txt

Error: Sorted input specified, but the file C_virginica-3.0_Gnomon_mRNA.gff3 has the following out of order record
NC_035780.1	Gnomon	mRNA	2413594	2416601	.	-	.	ID=rna199;Parent=gene122;Dbxref=GeneID:111129373,Genbank:XM_022475729.1;Name=XM_022475729.1;gbkey=mRNA;gene=LOC111129373;model_evidence=Supporting evidence includes similarity to: 2 Proteins;product=mucin-2-like;transcript_id=XM_022475729.1


The file was created, but the mRNA file itself is unsorted. I need to see if this impacted the output.

In [43]:
!head 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-Closest-NoOverlap-DMLs.txt

NC_035780.1	Gnomon	mRNA	28961	33324	.	+	.	ID=rna1;Parent=gene1;Dbxref=GeneID:111126949,Genbank:XM_022471938.1;Name=XM_022471938.1;gbkey=mRNA;gene=LOC111126949;model_evidence=Supporting evidence includes similarity to: 3 Proteins%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 21 samples with support for all annotated introns;product=UNC5C-like protein;transcript_id=XM_022471938.1	NC_035780.1	401630	401632	53	368307
NC_035780.1	Gnomon	mRNA	43111	66897	.	-	.	ID=rna2;Parent=gene2;Dbxref=GeneID:111110729,Genbank:XM_022447324.1;Name=XM_022447324.1;gbkey=mRNA;gene=LOC111110729;model_evidence=Supporting evidence includes similarity to: 1 Protein%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments;product=FMRFamide receptor-like%2C transcript variant X1;transcript_id=XM_022447324.1	NC_035780.1	401630	401632	53	334734
NC_035780.1	Gnomon	mRNA	43111	46506	.	-	.	ID=rna3;Parent=gene2;Dbxref=GeneID:111110729,Genbank:XM_022447333.1;Na

In [42]:
!tail 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-Closest-NoOverlap-DMLs.txt

NC_035780.1	Gnomon	mRNA	2028071	2046722	.	-	.	ID=rna177;Parent=gene107;Dbxref=GeneID:111121733,Genbank:XM_022463175.1;Name=XM_022463175.1;gbkey=mRNA;gene=LOC111121733;model_evidence=Supporting evidence includes similarity to: 4 ESTs%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 5 samples with support for all annotated introns;product=LIM and SH3 domain protein 1-like%2C transcript variant X6;transcript_id=XM_022463175.1	NC_035780.1	1933499	1933501	51	-94570
NC_035780.1	Gnomon	mRNA	2028071	2046721	.	-	.	ID=rna178;Parent=gene107;Dbxref=GeneID:111121733,Genbank:XM_022463184.1;Name=XM_022463184.1;gbkey=mRNA;gene=LOC111121733;model_evidence=Supporting evidence includes similarity to: 4 ESTs%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 5 samples with support for all annotated introns;product=LIM and SH3 domain protein 1-like%2C transcript variant X7;transcript_id=XM_022463184.1	NC_035780.1	1933499	193350

The output looks decent? I'll keep going.

In [47]:
! {bedtoolsDirectory}closestBed \
-io \
-a {mRNAList} \
-b {DMRlist} \
-t all \
-D ref \
> 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-Closest-NoOverlap-DMRs.txt

/bin/sh: {bedtoolsDirectory}closestBed: command not found


In [45]:
!head 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-Closest-NoOverlap-DMRs.txt

In [45]:
! {bedtoolsDirectory}closestBed \
-io \
-a {mRNAList} \
-b {CGMotifList} \
-t all \
-D ref \
> 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-Closest-NoOverlap-CGmotifs.txt

Error: Sorted input specified, but the file C_virginica-3.0_Gnomon_mRNA.gff3 has the following out of order record
NC_035780.1	Gnomon	mRNA	2413594	2416601	.	-	.	ID=rna199;Parent=gene122;Dbxref=GeneID:111129373,Genbank:XM_022475729.1;Name=XM_022475729.1;gbkey=mRNA;gene=LOC111129373;model_evidence=Supporting evidence includes similarity to: 2 Proteins;product=mucin-2-like;transcript_id=XM_022475729.1


In [46]:
!head 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-Closest-NoOverlap-CGmotifs.txt

NC_035780.1	Gnomon	mRNA	28961	33324	.	+	.	ID=rna1;Parent=gene1;Dbxref=GeneID:111126949,Genbank:XM_022471938.1;Name=XM_022471938.1;gbkey=mRNA;gene=LOC111126949;model_evidence=Supporting evidence includes similarity to: 3 Proteins%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments%2C including 21 samples with support for all annotated introns;product=UNC5C-like protein;transcript_id=XM_022471938.1	NC_035780.1	28924	28926	CG_motif	-35
NC_035780.1	Gnomon	mRNA	43111	66897	.	-	.	ID=rna2;Parent=gene2;Dbxref=GeneID:111110729,Genbank:XM_022447324.1;Name=XM_022447324.1;gbkey=mRNA;gene=LOC111110729;model_evidence=Supporting evidence includes similarity to: 1 Protein%2C and 100%25 coverage of the annotated genomic feature by RNAseq alignments;product=FMRFamide receptor-like%2C transcript variant X1;transcript_id=XM_022447324.1	NC_035780.1	66897	66899	CG_motif	1
NC_035780.1	Gnomon	mRNA	43111	46506	.	-	.	ID=rna3;Parent=gene2;Dbxref=GeneID:111110729,Genbank:XM_022447333.1;Na